[KM-86] error style

beckertobias opened the pull request:
https://kara-jobs.atlassian.net/browse/KM-86

- I put the error msg to layout. Because I want the error msg to be right below the header. If I didn't put it into the layout, it would be in `main`. And then I would have to use sth like `position=fixed` and set marginTop equal to the height of the header. I wanted to avoid this because position fixed is always a bit difficult on responsive websites. And also the height of the header is dynamic (with minHeight and maxHeight)
- you can close the error msg (similar to Github)

![Screenshot 2021-03-21 at 12 30 43](https://user-images.githubusercontent.com/63408795/111903184-43353580-8a41-11eb-82a4-8cf470723ebb.png)

TobiBecker committed:
style error component

TobiBecker committed:
change type of errorComponent

beckertobias committed:
Merge branch 'main' into KM-44-error-style

atsteinberg commented:
Suggestion (don’t know whether you’re already doing it that way: if so, cool!): Why don’t we structure layout as
Navbar
Error
Main
Footer

And then have components (in the case of main: dynamically chosen components depending on the page) to fill the layout. Wouldn’t that be the cleanest solution?

Since Apollo client doubles as state management we could set the error status through that and show the Error component depending on that global state.

As I said: maybe you’re already doing it this way. Maybe there are good reasons not to do it that way Currently, I can’t really look into the code. So, just a suggestion to consider.

> On 21. Mar 2021, at 12:31, Tobias Niklas Becker ***@***.***> wrote:
> 
> 
> @beckertobias <https://github.com/beckertobias> requested your review on: #43 <https://github.com/fullstacksmart/next-kara/pull/43> [KM-86] error style.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/43#event-4486628391>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIISR3ZFZKZJ7CRQO5F3TEXKJHANCNFSM4ZROW5YQ>.
>

atsteinberg commented:
But definitely looks cool!

> On 21. Mar 2021, at 12:31, Tobias Niklas Becker ***@***.***> wrote:
> 
> 
> @beckertobias <https://github.com/beckertobias> requested your review on: #43 <https://github.com/fullstacksmart/next-kara/pull/43> [KM-86] error style.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/43#event-4486628391>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIISR3ZFZKZJ7CRQO5F3TEXKJHANCNFSM4ZROW5YQ>.
>

atsteinberg commented:
Just one thing about the error message: I think it is considered best practice not to give out the information whether a user with an email address exists, but only that the email/password combination is invalid (for security reasons). Maybe we should check whether that’s the case.

> On 21. Mar 2021, at 12:31, Tobias Niklas Becker ***@***.***> wrote:
> 
> 
> @beckertobias <https://github.com/beckertobias> requested your review on: #43 <https://github.com/fullstacksmart/next-kara/pull/43> [KM-86] error style.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/43#event-4486628391>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIISR3ZFZKZJ7CRQO5F3TEXKJHANCNFSM4ZROW5YQ>.
>

TobiBecker committed:
adjust userNotFound translation

beckertobias commented:
good point regarding the userNotFound message. I changed the translations.

Regarding the other suggestions. I would say I am more or less already doing it this way:
Header
Error
Main
Footer

Error is a prop that is null by default. And I can pass this prop to layout:

<img width="417" alt="Screenshot 2021-03-21 at 15 01 43" src="https://user-images.githubusercontent.com/63408795/111907675-b1382780-8a56-11eb-9b33-acf7bd7fa266.png">

Not sure if we need to use a global state for this. Because the error messages that we want to show will only be relevant to the page where the user is currently on.

Maybe have a look at the code (when you have time) and then let me know what you think.

TobiBecker committed:
fix error component import on signin

atsteinberg commented on components/error/error-msg-styles.ts:
```diff
@@ -0,0 +1,11 @@
+import { makeStyles, createStyles } from '@material-ui/core/styles';
+
+const useStyles = makeStyles(({ spacing }) =>
```
you don't use spacing. Delete?

atsteinberg commented on containers/layout/Layout.tsx:
```diff
@@ -15,13 +15,15 @@ export interface LayoutProps {
   home?: boolean;
   heading?: string;
   title?: string | string[];
+  error: React.ReactNode | null;
```
Wouldn't it make sense to pass the error *message* here? We could then check whether it's an empty string (don't display) or not.

We would import the error component, passing it the error message if it gets displayed. Since we don't really change the whole component but only the message, this would make more sense to me. What do you think?

beckertobias replied:
That makes sense. It's a good suggestion.
I changed it. So I am passing the error of type `FirebaseError` (this error has the `code` and `message` that I need) instead of passing the error component.

atsteinberg commented on pages/signin/index.tsx:
```diff
@@ -43,9 +43,10 @@ const SignInPage = ({ t }: PageProps): React.ReactElement => {
     }
   };
 
+  const errorComponent = error ? <Error error={error} /> : null;
```
see above. I think we should not create the errorComponent here and pass it to layout, but pass the message and import the component into layout.

atsteinberg commented on pages/signin/index.tsx:
```diff
@@ -43,9 +43,10 @@ const SignInPage = ({ t }: PageProps): React.ReactElement => {
   return (
-    <Layout title="sign in" t={t}>
-      {error && <Error error={error} />}
+    <Layout title="sign in" t={t} error={errorComponent}>
```
see above

atsteinberg reviewed (commented):
Very nice. I have one suggestion for slightly changing how things are set up, but very cool over all.

One thing I noticed. I think the files we define our components in should be named like the component (e.g. Error.tsx), not index.tsx (index.ts should only be used to import/export the component. Otherwise all of our components live in files with the same name. We can talk about this, if you don't agree, but too me it would make sense. (right now we have a mixed bag, I think). For pages as well, actually.

atsteinberg commented:
@beckertobias could you have a look at the comments, please. This is almost ready to be merged.

TobiBecker committed:
remove spacing

TobiBecker committed:
pass firebase error instead of component

beckertobias commented:
@atsteinberg I made some changes based on your feedback. Please approve if fine for you.

Regarding your comment on the naming convention of the files. I will create a jira ticket so that we agree on these conventions and then stick to them. We currently have mixed conventions.

Ticket: https://kara-jobs.atlassian.net/browse/KM-97

TobiBecker committed:
basic pw validaiton

beckertobias commented:
Okay, I am adding the password validation here now as well to avoid conflicts. 
PR is WORK IN PROGRESS for now.

TobiBecker committed:
refactor handleSubmit

TobiBecker committed:
cleanup

beckertobias committed:
Merge branch 'main' into KM-44-error-style

TobiBecker committed:
make sure error disappears when entering again

TobiBecker committed:
merge main

Merge branch 'KM-44-error-style' of https://github.com/fullstacksmart/next-kara into KM-44-error-style

beckertobias commented:
@atsteinberg okay now it's ready for review and also contains the password validation.

I'm not completely happy with it. I think it would have been easier and more neat if we had used Formik from the beginning on: https://formik.org/docs/examples/with-material-ui. But if we used that now, we would have to refactor quite a lot. Also we have a custom `InputField`. We would have to check if we can even do that with a custom `InputField`.

And I did the check that I `console.log` something to see how often we update the component (because of all the `useStates`. But actually I only get the logging once if I update sth. So I didn't see the need to use the `useReducer`.

beckertobias commented on pages/signup/index.tsx:
```diff
@@ -39,22 +39,46 @@ const ADD_TALENT = gql`
+      const newValue = e.target.value;
+      return {
+        ...oldValues,
+        ...computeNestedValue(oldValues, propArray, newValue),
```
I am getting a type error in line 77 because the func `computeNestedValue` wants the type for the first argument: Record<string, unknown>. Do you have an idea how to solve this one?

`Argument of type 'SignupFormValues' is not assignable to parameter of type 'Record<string, unknown>'.
  Type 'NewBaseUser & { type: UserType; }' is not assignable to type 'Record<string, unknown>'.
    Index signature is missing in type 'NewBaseUser & { type: UserType; }'.`

atsteinberg replied:
The problem is with using Record<string, unknown> with an interface.: https://github.com/microsoft/TypeScript/issues/15300

It can be fixed by making SignupFormValues into a type. If you replace the end of the types file with this:
```

export type BaseUser = {
  gender: Gender;
  name: FullName;
  email: string;
};

type NewBaseUser = BaseUser & {
  password: string;
};

export type NewTalent = NewBaseUser;

export type NewEmployer = NewBaseUser & {
  company: string;
};

export type SignupFormValues = (NewTalent | NewEmployer) & {
  type: UserType;
};
```
that should fix the issue

TobiBecker committed:
absolute import

TobiBecker committed:
more absolute imports

TobiBecker committed:
small cleanup

TobiBecker committed:
no .env

atsteinberg commented:
@TobiBecker I agree that we should have used a form helper package. Personally I don't like Formik all that much, but react-hook-form has similar functionality and produces code that is much easier to read. I actually considered suggesting to introduce it but then ran into the same issue that it would be hell to replace our existing code (on the face of it). Maybe we could look into this for a refactor opportunity before adding the employer profile.

atsteinberg reviewed (changes requested):
If you update the types, this should be good to go

TobiBecker committed:
fix type issue

beckertobias commented:
Nice, thanks for helping out with the type issue.

And regarding the form helper package: Yes, sounds like a plan. Before we implement the employer profile we can look into this.

atsteinberg reviewed (approved):
👍

TobiBecker committed:
[KM-89] Reactive variables: Don't load layout on every page (#49)

* trying out reactive variables

* useReactiveVar

* basic setup working

* implementation for signin

* Fix types

* small cleanup

* cleanup

* define variables in apollo/cache

* load basic info in layout

* small refactor

* refacotor imports profile page

* remove layout from all pages

* small cleanup

* fix initial type layoutError

* rename reactiveVar

* isOnTalentPage

* >isOnTalentProfilePage

* remove 404 page layout

* Landing > Home

* new layoutVar

* changes in layout container

* separate layoutVar

* layout container with separate layoutVar

* set title heading in profile component

* upgrade getTitleStringFromPathname

* cleanup

* more cleanup

TobiBecker committed:
merge master into branch

TobiBecker committed:
fix error after merging
